DRA: test plugin re-registration after failed GetInfo call

bart0sh opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Implemented e2e_node test for this use case:

If `GetInfo` call fails, it's expected to be called again by the kubelet plugin manager. This allows to complete plugin registration after plugin is able to successfully process the call.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

bart0sh committed:
e2e: dra: simulate GetInfo failure

bart0sh committed:
e2e: dra: implement ResetGRPCCalls

bart0sh committed:
e2e_node: dra: test plugin registration retry

Adds a DRA e2e_node test to verify that the kubelet plugin manager
retries plugin registration when the GetInfo call fails, and
successfully registers the plugin once GetInfo succeeds.

This ensures correct recovery and registration behavior for
DRA plugins in failure scenarios.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -663,6 +663,10 @@ func (d *Helper) RegistrationStatus() *registerapi.RegistrationStatus {
 	return d.registrar.status
 }
 
+func (d *Helper) SetGetInfoFailureMode() func() {
```
Docs?

bart0sh replied:
done

pohly commented on test/e2e/dra/test-driver/app/gomega.go:
```diff
@@ -33,6 +33,15 @@ var BeRegistered = gcustom.MakeMatcher(func(actualCalls []GRPCCall) (bool, error
+	}
+	return false, nil
+}).WithMessage("contain unsuccessful GetInfo call")
+
```
Can we use this opportunity to move the gomega matchers into a `test/e2e/dra/test-driver/testing` package which then gets imported as `testdrivertesting`?

They don't belong here and the usage as `testdriver.GetInfoFailed` looked confusing until I realised that this is a matcher, not some function.

Alternatively, `gomega` instead of `testing` also works. It would avoid the `test` + `testing` repetition.

Side note: it might be better to turn `GetInfoFailed` into a function which returns a new instance. Some Gomega matchers store information, which can become confusing when the same instance gets used multiple times. I don't know whether that is a problem here, but let's be on the safe side.

bart0sh replied:
Can I do it in a separate PR?

bart0sh replied:
After this one is merged :)

pohly replied:
Why not now? Then we don't need to bother approvers twice.

bart0sh replied:
> Why not now?

Because it's out of scope of this PR?

pohly replied:
Adding `GetInfoFailed` here is not good, so doing something better is in scope for this PR.

If you feel that a separate cleanup PR is necessary, then do that first and we wait with this PR until the cleanup PR is merged.

pohly replied:
But I think a single PR with several commits would be less work for everyone involved. Up to you.

bart0sh replied:
Thank you for the review!
I've updated this PR as you've suggested. PTAL

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -663,6 +663,12 @@ func (d *Helper) RegistrationStatus() *registerapi.RegistrationStatus {
 }
 
+// SetGetInfoFailureMode configures the registration server to make
+// GetInfo calls fail. Returns a function that restores normal behavior.
```
Fail how?

bart0sh replied:
Updated, PTAL.

bart0sh committed:
e2e: dra: move gomega matchers to dedicated package

Moved gomega matcher definitions from test-driver/app
to a new test-driver/gomega package.

SergeyKanzhelev commented on test/e2e_node/dra_test.go:
```diff
@@ -123,7 +124,30 @@ var _ = framework.SIGDescribe("node")(framework.WithLabel("DRA"), feature.Dynami
+			unsetGetInfoFailureMode()
+
+			ginkgo.By("wait for Kubelet plugin re-registration")
+			gomega.Eventually(kubeletPlugin.GetGRPCCalls).WithTimeout(pluginRegistrationTimeout).Should(testdrivergomega.BeRegistered)
```
curious what is the actual reconnect speed? In case of a device plugin we typically implement it to try to reconnect as fast as it can. So we are not failing Pods that are already scheduled on the node after the node restart.

bart0sh replied:
This is what I can see in the Kubelet logs when it tries to reconnect:

```
I0515 19:45:47.606393 3556532 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-8372.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:47.606238972+03:00","UUID":"41f85631-f078-44c0-8c51-1220989f1c1c","Handler":null,"Name":"","Endpoint":""}
I0515 19:45:47.606568 3556532 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1021.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:47.606085411+03:00","UUID":"957f7cd6-0890-4a07-bf88-77633d47aa3f","Handler":null,"Name":"","Endpoint":""}
I0515 19:45:47.606610 3556532 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1782.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:47.606166055+03:00","UUID":"e7ab2015-d5a7-4b5e-a837-fffe5ff73119","Handler":null,"Name":"","Endpoint":""}
I0515 19:45:47.606646 3556532 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-7779.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:47.606201878+03:00","UUID":"6a0332dd-c460-4df0-b65e-a1398f0f2658","Handler":null,"Name":"","Endpoint":""}
I0515 19:45:48.607381 3556532 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/test-driver.cdi.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:48.516662594+03:00","UUID":"375245f3-0e32-46c8-a820-bb169acbe257","Handler":null,"Name":"","Endpoint":""}
I0515 19:45:49.849255 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1021.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848838295+03:00","UUID":"c09a5d3f-b4dc-4173-9fab-443c850cbfa0","Handler":null,"Name":"","Endpoint":""}
I0515 19:45:49.849395 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1782.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848886458+03:00","UUID":"782c1b65-d8b9-4229-aa2e-d84e8b162f8d","Handler":null,"Name":"","Endpoint":""}
I0515 19:45:49.849419 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-7779.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848936355+03:00","UUID":"44d9fffa-bf70-48a1-a92f-dfa02300efe3","Handler":null,"Name":"","Endpoint":""}
I0515 19:45:49.849439 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-8372.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848991855+03:00","UUID":"f43236f8-44ea-46d1-bd0f-884f9e8671ca","Handler":null,"Name":"","Endpoint":""}
I0515 19:45:49.849458 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/test-driver.cdi.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.84905169+03:00","UUID":"038874c9-f4e4-4b5c-b8aa-f030e998d21a","Handler":null,"Name":"","Endpoint":""}
I0515 19:45:51.851671 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/test-driver.cdi.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.84905169+03:00","UUID":"038874c9-f4e4-4b5c-b8aa-f030e998d21a","Handler":null,"Name":"","Endpoint":""}
I0515 19:45:54.853781 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/test-driver.cdi.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.84905169+03:00","UUID":"038874c9-f4e4-4b5c-b8aa-f030e998d21a","Handler":null,"Name":"","Endpoint":""}
I0515 19:45:57.856294 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/test-driver.cdi.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.84905169+03:00","UUID":"038874c9-f4e4-4b5c-b8aa-f030e998d21a","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:00.858694 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-7779.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848936355+03:00","UUID":"44d9fffa-bf70-48a1-a92f-dfa02300efe3","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:00.858758 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-8372.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848991855+03:00","UUID":"f43236f8-44ea-46d1-bd0f-884f9e8671ca","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:00.858789 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1021.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848838295+03:00","UUID":"c09a5d3f-b4dc-4173-9fab-443c850cbfa0","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:00.858811 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1782.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848886458+03:00","UUID":"782c1b65-d8b9-4229-aa2e-d84e8b162f8d","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:02.860266 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/test-driver.cdi.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.84905169+03:00","UUID":"038874c9-f4e4-4b5c-b8aa-f030e998d21a","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:11.864081 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1021.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848838295+03:00","UUID":"c09a5d3f-b4dc-4173-9fab-443c850cbfa0","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:11.864236 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1782.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848886458+03:00","UUID":"782c1b65-d8b9-4229-aa2e-d84e8b162f8d","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:11.864263 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-7779.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848936355+03:00","UUID":"44d9fffa-bf70-48a1-a92f-dfa02300efe3","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:11.864282 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-8372.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848991855+03:00","UUID":"f43236f8-44ea-46d1-bd0f-884f9e8671ca","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:11.864301 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/test-driver.cdi.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.84905169+03:00","UUID":"038874c9-f4e4-4b5c-b8aa-f030e998d21a","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:23.872419 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1021.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848838295+03:00","UUID":"c09a5d3f-b4dc-4173-9fab-443c850cbfa0","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:23.872554 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1782.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848886458+03:00","UUID":"782c1b65-d8b9-4229-aa2e-d84e8b162f8d","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:23.872583 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-7779.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848936355+03:00","UUID":"44d9fffa-bf70-48a1-a92f-dfa02300efe3","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:23.872626 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-8372.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848991855+03:00","UUID":"f43236f8-44ea-46d1-bd0f-884f9e8671ca","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:28.875122 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/test-driver.cdi.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.84905169+03:00","UUID":"038874c9-f4e4-4b5c-b8aa-f030e998d21a","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:37.880739 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1021.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848838295+03:00","UUID":"c09a5d3f-b4dc-4173-9fab-443c850cbfa0","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:37.880986 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1782.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848886458+03:00","UUID":"782c1b65-d8b9-4229-aa2e-d84e8b162f8d","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:37.881021 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-7779.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848936355+03:00","UUID":"44d9fffa-bf70-48a1-a92f-dfa02300efe3","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:37.881040 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-8372.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848991855+03:00","UUID":"f43236f8-44ea-46d1-bd0f-884f9e8671ca","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:55.893040 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1021.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848838295+03:00","UUID":"c09a5d3f-b4dc-4173-9fab-443c850cbfa0","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:55.893200 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1782.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848886458+03:00","UUID":"782c1b65-d8b9-4229-aa2e-d84e8b162f8d","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:55.893236 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-7779.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848936355+03:00","UUID":"44d9fffa-bf70-48a1-a92f-dfa02300efe3","Handler":null,"Name":"","Endpoint":""}
I0515 19:46:55.893256 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-8372.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848991855+03:00","UUID":"f43236f8-44ea-46d1-bd0f-884f9e8671ca","Handler":null,"Name":"","Endpoint":""}
I0515 19:47:01.895846 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/test-driver.cdi.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.84905169+03:00","UUID":"038874c9-f4e4-4b5c-b8aa-f030e998d21a","Handler":null,"Name":"","Endpoint":""}
I0515 19:47:21.908304 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1021.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848838295+03:00","UUID":"c09a5d3f-b4dc-4173-9fab-443c850cbfa0","Handler":null,"Name":"","Endpoint":""}
I0515 19:47:21.908442 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1782.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848886458+03:00","UUID":"782c1b65-d8b9-4229-aa2e-d84e8b162f8d","Handler":null,"Name":"","Endpoint":""}
I0515 19:47:21.908482 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-7779.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848936355+03:00","UUID":"44d9fffa-bf70-48a1-a92f-dfa02300efe3","Handler":null,"Name":"","Endpoint":""}
I0515 19:47:21.908508 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-8372.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848991855+03:00","UUID":"f43236f8-44ea-46d1-bd0f-884f9e8671ca","Handler":null,"Name":"","Endpoint":""}
I0515 19:48:03.934613 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-7779.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848936355+03:00","UUID":"44d9fffa-bf70-48a1-a92f-dfa02300efe3","Handler":null,"Name":"","Endpoint":""}
I0515 19:48:03.934761 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-8372.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848991855+03:00","UUID":"f43236f8-44ea-46d1-bd0f-884f9e8671ca","Handler":null,"Name":"","Endpoint":""}
I0515 19:48:03.934801 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1021.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848838295+03:00","UUID":"c09a5d3f-b4dc-4173-9fab-443c850cbfa0","Handler":null,"Name":"","Endpoint":""}
I0515 19:48:03.934860 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1782.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848886458+03:00","UUID":"782c1b65-d8b9-4229-aa2e-d84e8b162f8d","Handler":null,"Name":"","Endpoint":""}
I0515 19:48:06.936678 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/test-driver.cdi.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.84905169+03:00","UUID":"038874c9-f4e4-4b5c-b8aa-f030e998d21a","Handler":null,"Name":"","Endpoint":""}
I0515 19:49:17.980217 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1021.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848838295+03:00","UUID":"c09a5d3f-b4dc-4173-9fab-443c850cbfa0","Handler":null,"Name":"","Endpoint":""}
I0515 19:49:17.980382 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-1782.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848886458+03:00","UUID":"782c1b65-d8b9-4229-aa2e-d84e8b162f8d","Handler":null,"Name":"","Endpoint":""}
I0515 19:49:17.980414 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-7779.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848936355+03:00","UUID":"44d9fffa-bf70-48a1-a92f-dfa02300efe3","Handler":null,"Name":"","Endpoint":""}
I0515 19:49:17.980435 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/dra-8372.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.848991855+03:00","UUID":"f43236f8-44ea-46d1-bd0f-884f9e8671ca","Handler":null,"Name":"","Endpoint":""}
I0515 19:50:10.012376 3556707 reconciler.go:161] "OperationExecutor.RegisterPlugin started" plugin={"SocketPath":"/var/lib/kubelet/plugins_registry/test-driver.cdi.k8s.io-reg.sock","Timestamp":"2025-05-15T19:45:49.84905169+03:00","UUID":"038874c9-f4e4-4b5c-b8aa-f030e998d21a","Handler":null,"Name":"","Endpoint":""}
```

bart0sh commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/registrationserver.go:
```diff
@@ -29,12 +31,18 @@ type registrationServer struct {
 // GetInfo is the RPC invoked by plugin watcher.
 func (e *registrationServer) GetInfo(ctx context.Context, req *registerapi.InfoRequest) (*registerapi.PluginInfo, error) {
+	if failure := e.getGetInfoFailure(); failure != nil {
+		return nil, failure
```
> is there a difference if this is a failure or timeout?

I can only say that the test succeeded when I changed the code to sleep for 2 min instead of returning an error.
Reconcilation timeout of the plugin manager is 1s, so it 

Thanks for the idea, I'll come up with new test case(s) simulating timeout, but not in this PR.

> Also, it is a good idea to document a best practice for the plugin writers. Should it return the endpoint address even before it listens on it or it should ensure it listens on the port first and then return success from GetInfo? What will be the difference in bootstrap time in these two cases?

Absolutely. We should create/update documentation for plugin writers. What would be a good place for it in your opinion? I'd also propose to check how it's implemented in the [DRA example driver](https://github.com/kubernetes-sigs/dra-example-driver) and change it if needed.

/cc @lauralorenz @nojnhuh

SergeyKanzhelev replied:
> I can only say that the test succeeded when I changed the code to sleep for 2 min instead of returning an error.
Reconcilation timeout of the plugin manager is 1s, so it

So it also ensures that the timeout is `1s` from the kubelet side? Or you have also changed the registration timeout in the test? That may be a good thing to confirm if you plan to add timeout tests.

> We should create/update documentation for plugin writers.

@lauralorenz can you please adda task to the board?

bart0sh replied:
> So it also ensures that the timeout is 1s from the kubelet side?

In fact it's less than that. See the log output in another comment below.
However, the reconcilation timeout is 1s: it's defined [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pluginmanager/plugin_manager.go#L49) and used [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pluginmanager/reconciler/reconciler.go#L85-L91) if I read the code correctly.

> Or you have also changed the registration timeout in the test?

No, I didn't. I only changed it in the test plugin's `GetInfo` to 150 sec:
```
@@ -41,7 +42,8 @@ var _ registerapi.RegistrationServer = &registrationServer{}
 // GetInfo is the RPC invoked by plugin watcher.
 func (e *registrationServer) GetInfo(ctx context.Context, req *registerapi.InfoRequest) (*registerapi.PluginInfo, error) {
        if failure := e.getGetInfoFailure(); failure != nil {
-               return nil, failure
+               time.Sleep(time.Second * 150)
+               //return nil, failure
        }
```

and changed the test:
```
diff --git a/test/e2e_node/dra_test.go b/test/e2e_node/dra_test.go
index 0c819427ce5..4a8b8d1a52d 100644
--- a/test/e2e_node/dra_test.go
+++ b/test/e2e_node/dra_test.go
@@ -140,8 +140,9 @@ var _ = framework.SIGDescribe("node")(framework.WithLabel("DRA"), feature.Dynami
                        restartKubelet(ctx, true)
 
                        ginkgo.By("wait for Registration call to fail")
-                       gomega.Eventually(kubeletPlugin.GetGRPCCalls).WithTimeout(retryTestTimeout).Should(testdrivergomega.GetInfoFailed())
-                       gomega.Expect(kubeletPlugin.GetGRPCCalls()).ShouldNot(testdrivergomega.BeRegistered, "Expect plugin not to be registered due to GetInfo failure")
+                       //gomega.Eventually(kubeletPlugin.GetGRPCCalls).WithTimeout(retryTestTimeout).Should(testdrivergomega.GetInfoFailed())
+                       //gomega.Expect(kubeletPlugin.GetGRPCCalls()).ShouldNot(testdrivergomega.BeRegistered, "Expect plugin not to be registered due to GetInfo failure")
+                       gomega.Consistently(kubeletPlugin.GetGRPCCalls).WithTimeout(2*retryTestTimeout).ShouldNot(testdrivergomega.BeRegistered, "Expect plugin not to be registered due to GetInfo timeout")
 
                        ginkgo.By("unset registration failure mode")
                        unsetGetInfoFailureMode()
```

pohly replied:
Our example driver uses it, and we point developers/plugin authors to that.

We don't have a dedicated "developer documentation" page.

bart0sh replied:
Sounds good enough for me. Anyway, this is not related to this PR, so can you lgtm/approve it? The documentation issue(if any) should be handled separately as @SergeyKanzhelev suggested above.

SergeyKanzhelev replied:
> We don't have a dedicated "developer documentation" page.

I wonder if this will be needed. As we unlikely will cover all those cases with tests. And writing considerations in English sounds very useful. Something like:

1. GetInfo MUST not perform I/O operation as the timeout is small and any unexpected latency on those operations will result in plugin never registering.
2. GetInfo MUST return error quickly if the endpoint is not ready. It MUST not implement any lock to wait for the endpoint to be ready. This will allow other plugins to register faster as plugins are being registered serially.
3. GetInfo MUST return OK only when endpoint is ready. So in the case of upgrade, kubelet will only switch to the new endpoint when it is ready.

Then we can cover good practices of liveness, and startup probes.

bart0sh commented:
/retest

bart0sh commented:
/retest

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/registrationserver.go:
```diff
@@ -29,12 +31,18 @@ type registrationServer struct {
 	supportedVersions []string
 	status            *registerapi.RegistrationStatus
+
+	getInfoFailure   error
+	failGetInfoMutex sync.Mutex
```
For single-field atomic access, atomic.Pointer is simpler. It also prevents accidentally accessing the field without the mutex being locked:
```suggestion
	getInfoError   atomic.Pointer[error]
```

bart0sh replied:
Done, PTAL and approve, thanks!

bart0sh commented on staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go:
```diff
@@ -663,6 +663,13 @@ func (d *Helper) RegistrationStatus() *registerapi.RegistrationStatus {
 }
 
+// SetGetInfoFailureMode configures the registration server to make
+// GetInfo calls return an error.
```
done, thanks for the suggestion!

bart0sh commented:
/test pull-kubernetes-e2e-kind

pohly reviewed (approved):
/lgtm
